### ISSUE TITLE
interleaved log lines during parallel operations

### ISSUE THREAD

**Issue by terrorobe** [pgmasters/backrest#97]
Currently it's possible to have interleaved log writes when doing parallel backups (and possibly restores as well) of many small files:

```
2015-06-02 03:00:05.897 T01   INFO: backup file /data/postgresql/9.3/main/base/12078/11999 (16KB) checksum d8191f66213378f201c9a972e575e63cb48b8140
2015-06-02 03:00:05.899 T01   INFO: backup file /data/postgresql/9.3/main/base/12078/12000_vm (8KB) checksum c7e1ca68e3e1da2e33a0798f4dfab8d9c7e1101e
2015-06-02 03:00:05.900 T01   INFO: backup file /data/postgresql/9.3/main/base/12078/12003 (16KB) checksum b0e254f511eb80bff768e8fccb03e7342362aa8b
2015-06-02 03:00:05.902 T01   INFO: backup file /data/postgresql/9.3/main/base/12078/12006 (8KB) checksu: backup file /data/postgresql/9.3/main/base/12078/11941_vm (8KB) checksum b416ab82ddac10e6a8e01e077d7b045c0fdc70e6
2015-06-02 03:00:05.858 T02   INFO: backup file /data/postgresql/9.3/main/base/12078/11944 (16KB) checksum 1485f25667c8fe8ff1e63930ba386405ae9da2ab
2015-06-02 03:00:05.873 T02   INFO: backup file /data/postgresql/9.3/main/base/12078/11946 (376KB) checksum 209b31fc37e67e36e4107c9f922738a3ebe6309f
2015-06-02 03:00:05.884 T02   INFO: backup file /data/postgresql/9.3/main/base/12078/11949 (224KB) checksum ee48c6342dc79daf0415d646462c77f4960a9b0e

[..]

2015-06-02 03:00:05.973 T02   INFO: backup file /data/postgresql/9.3/main/base/16385/11828 (8KB) checksum ea2b4b8f95bd3f06c8d99f6a36a424e72c1cd501
2015-06-02 03:00:05.974 T02   INFO: backup file /data/postgresql/9.3/main/base/16385/11829_fsm (24KB) checksum d804ee2797cd73f276ca7fa51359d7f2fd73e28b
2015-06-02 03:00:05.975 T02   INFO: backup file /data/postgresql/9.3/main/base/16385/11829_vm (8KB) checksum 56f5c506c95eb15a439aadf5f1fd0306edd96e16
2015-06-02 03:00:05.989 T02   INFO: backm ea2b4b8f95bd3f06c8d99f6a36a424e72c1cd501
2015-06-02 03:00:05.903 T01   INFO: backup file /data/postgresql/9.3/main/base/12078/12008 (0B)
2015-06-02 03:00:05.904 T01   INFO: backup file /data/postgresql/9.3/main/base/12078/12011 (8KB) checksum ea2b4b8f95bd3f06c8d99f6a36a424e72c1cd501
2015-06-02 03:00:05.905 T01   INFO: backup file /data/postgresql/9.3/main/base/12078/12014 (8KB) checksum ea2b4b8f95bd3f06c8d99f6a36a424e72c1cd501
```

Complete log: http://nopaste.narf.at/show/MdKxgWwuudDvENsq7zKX/

No dealbreaker, more of a cosmetic thing, probably good to address this with #48

**Comment by terrorobe** [comment]
On second thought, the large buffering of the main thread can have some very distorting effects, e.g.:


```
2015-06-02 09:00:02.802 T00   INFO: Restoring backup set latest
2015-06-02 09:00:03.178 T00   WARN: group root cannot be set, changed to postgres
2015-06-02 09:00:03.182 T00   WARN: user root cannot be set, changed to postgres
2015-06-02 09:00:03.182 T00   INFO: checking/cleaning db path /data/postgresql/9.3/main
2015-06-02 09:00:03.242 T00   INFO: removing file/link /data/postgresql/9.3/main/pg_xlog/archive_status/0000000100001EFC0000007C.done
2015-06-02 09:00:03.242 T00   INFO: removing file/link /data/postgresql/9.3/main/pg_xlog/archive_status/0000000100001EFC0000007B.done
2015-06-02 09:00:03.242 T00   INFO: removing file/link /data/postgresql/9.3/main/pg_xlog/archive_status/0000000100001EFC0000007A.done
2015-06-02 09:00:03.242 T00   INFO: removing file/link /data/postgresql/9.3/main/pg_xlog/archive_status/0000000100001EFC00000079.done
2015-06-02 09:00:03.242 T00   INFO: removing file/link /data/postgresql/9.3/main/pg_xlog/archive_status/0000000100001EFC00000078.done
2015-06-02 09:00:03.242 T00   INFO: removing file/link /data/postgresql/9.3/main/pg_xlog/archive_status/0000000100001EFC00000077.done
2015-06-02 09:00:03.243 T00   INFO: removing file/link /data/postgresql/9.3/main/pg_xlog/archive_status/0000000100001EFC00000076.done

[..]

2015-06-02 09:00:03.309 T00   INFO: removing file/link /data/postgresql/9.3/main/pg_xlog/0000000100001EFB0000004C
2015-06-02 09:00:03.309 T00   INFO: removing file/link /data/postgresql/9.3/main/pg_xl2015-06-02 09:00:28.934 T01   INFO: /data/postgresql/9.3/main/base/16668/1732552149.5 restore (1GB) checksum 1da3f1d0998ac3c213293520d03bb331d1dd66b1
2015-06-02 09:00:49.274 T01   INFO: /data/postgresql/9.3/main/base/16668/1732552149.3 restore (1GB) checksum 479dc2d5650cbf2606383224354c5a6cbe317019
2015-06-02 09:01:12.043 T01   INFO: /data/postgresql/9.3/main/base/16668/1732552149.1 restore (1GB) checksum 69337b9125446c0a9f67eb3f36028cfc220869ff
2015-06-02 09:01:31.532 T01   INFO: /data/postgresql/9.3/main/base/16385/1871505399.9 restore (1GB) checksum 0f9d61da1dcbd04954accfb63e7dc35a21b9a1b8
2015-06-02 09:01:50.305 T01   INFO: /data/postgresql/9.3/main/base/16385/1871505399.7 restore (1GB) checksum 920a5acf8939d8a4bf01c761899c0c65ddfd5ab6
2015-06-02 09:02:10.647 T01   INFO: /data/postgresql/9.3/main/base/16385/1871505399.5 restore (1GB) checksum 6bc6265b51df0dd0faa9b7e54604bb7d157c5f95

[..]


2015-06-02 09:03:47.341 T02   INFO: /data/postgresql/9.3/main/base/1/11846 restore (0B)
2015-06-02 09:03:47.342 T02   INFO: /data/postgresql/9.3/main/base/1/11835 restore (0B)
2015-06-02 09:03:47.343 T02   INFO: /data/postgresql/9.3/main/base/1/11825 restore (0B)
og/0000000100001EFB0000004B
2015-06-02 09:00:03.309 T00   INFO: removing file/link /data/postgresql/9.3/main/pg_xlog/0000000100001EFB0000004A
2015-06-02 09:00:03.309 T00   INFO: removing file/link /data/postgresql/9.3/main/pg_xlog/0000000100001EFB00000049
2015-06-02 09:00:03.309 T00   INFO: removing file/link /data/postgresql/9.3/main/pg_xlog/0000000100001EFB00000048
2015-06-02 09:00:03.309 T00   INFO: removing file/link /data/postgresql/9.3/main/pg_xlog/0000000100001EFB00000047
2015-06-02 09:00:03.309 T00   INFO: removing file/link /data/postgresql/9.3/main/pg_subtrans/AE33
2015-06-02 09:00:03.309 T00   INFO: removing file/link /data/postgresql/9.3/main/pg_stat/global.stat


[..]

2015-06-02 09:00:03.327 T00   INFO: removing file/link /data/postgresql/9.3/main/base/16385/1865457019_vm
2015-06-02 09:00:03.327 T00   INFO: removing file/link /data/postgresql/9.3/main/base/16385/1865457019_fsm
2015-06-02 09:00:03.327 T00   INFO: removing file/link /data/postgresql/9.3/main/base/16385/1865457019
2015-06-02 09:00:03.338 T00   INFO: 639 file(s) removed during cleanup
2015-06-02 09:00:03.338 T00   INFO: 1 path(s) removed during cleanup
2015-06-02 09:03:47.634 T00   INFO: restore complete
```

Complete log: http://nopaste.narf.at/show/ILhyZoG9dUFC5fGCy2yB/

**Comment by dwsteele** [comment]
I completely agree that this can be annoying - and is more than cosmetic.  However, the issue will completely go away when moving to processes and it doesn't seem worth trying to fix this in the threaded version.  I did figure out how to make it work but it's not trivial.

**Comment by terrorobe** [comment]
:+1: - let's just keep this around for documentation.

Though I was wondering if it wouldn't suffice to move to unbuffered output for the log FHs - haven't encountered any issues on console output so far. But yeah, busywork ;)

**Comment by dwsteele** [comment]
What a perfectly prosaic solution.  Turns out it works pretty well.

Fixed in af98f16.
